src/calculation/CompositeReportBuilder.ts

```diff
@@ -70,6 +89,71 @@ export class CompositeReportBuilder<T extends PopulationGroupResult> extends Abs
     };
   }
 
+  public addResults(results: ExecutionResult<T>[]) {
+    // the weighted composite scoring type is component-based, but the others are individual-based
+    if (this.compositeScoringType === 'weighted') {
+      this.addComponentResults(results);
+    } else {
+      results.forEach(result => {
+        this.addPatientResults(result);
+      });
+    }
+  }
+
+  private addComponentResults(results: ExecutionResult<T>[]) {
+    // https://build.fhir.org/ig/HL7/cqf-measures/composite-measures.html#weighted-scoring
+    const componentPopulationResults: Record<string, { numerator: number; denominator: number; weight: number }> = {};
+
+    results.forEach(result => {
+      result.componentResults?.forEach(componentResult => {
+        if (componentResult.componentCanonical && !componentPopulationResults[componentResult.componentCanonical]) {
+          componentPopulationResults[componentResult.componentCanonical] = {
+            numerator: 0,
+            denominator: 0,
+            weight: this.weightedComponents[componentResult.componentCanonical]
+          };
+        }
+
+        if (componentResult.componentCanonical && componentPopulationResults[componentResult.componentCanonical]) {
+          if (
+            componentResult.populationResults?.find(pr => pr.populationType === PopulationType.NUMER)?.result === true
+          ) {
+            componentPopulationResults[componentResult.componentCanonical].numerator++;
+          }
+
+          if (
+            componentResult.populationResults?.find(pr => pr.populationType === PopulationType.DENOM)?.result === true
+          ) {
+            componentPopulationResults[componentResult.componentCanonical].denominator++;
+          }
+        }
+      });
+
+      if (result.evaluatedResource) {
+        result.evaluatedResource.forEach(resource => {
+          const reference: fhir4.Reference = {
+            reference: `${resource.resourceType}/${resource.id}`
+          };
+          if (!this.report.evaluatedResource?.some(r => r.reference === reference.reference)) {
+            if (!this.report.evaluatedResource) {
+              this.report.evaluatedResource = [reference];
+            } else {
+              this.report.evaluatedResource.push(reference);
+            }
+          }
+        });
+      }
+    });
+
+    // Multiply each of the component ratios by the weight associated with the component
+    // TODO: this may need to change if an extension is created for non-integer counts on the measure report
+    Object.values(componentPopulationResults).forEach(value => {
+      this.compositeFraction.numerator += value.weight * (value.numerator / value.denominator);
```

Comment:
I think we should guard against divide by 0 here. So if `value.denominator` is 0 then we shouldn't increment the numerator, or ensure that it gets incremented by 0